pkg/beacon/relay/event/deduplicator.go

```diff
@@ -24,54 +24,22 @@ type chain interface {
 // should be handled.
 //
 // Those events are supported:
-// - group selection started
 // - relay entry requested
 type Deduplicator struct {
-	chain                          chain
-	minGroupCreationDurationBlocks uint64
-
-	groupSelectionMutex             sync.Mutex
-	currentGroupSelectionStartBlock uint64
+	chain chain
 
 	relayEntryMutex             sync.Mutex
 	currentRequestStartBlock    uint64
 	currentRequestPreviousEntry string
 }
 
 // NewDeduplicator constructs a new Deduplicator instance.
-func NewDeduplicator(
-	chain chain,
-	minGroupCreationDurationBlocks uint64,
-) *Deduplicator {
+func NewDeduplicator(chain chain) *Deduplicator {
 	return &Deduplicator{
-		chain:                          chain,
-		minGroupCreationDurationBlocks: minGroupCreationDurationBlocks,
+		chain: chain,
 	}
 }
 
-// NotifyGroupSelectionStarted notifies the client wants to start group
-// selection upon receiving an event. It returns boolean indicating whether the
-// client should proceed with the execution or ignore the event as a duplicate.
-func (d *Deduplicator) NotifyGroupSelectionStarted(
-	newGroupSelectionStartBlock uint64,
-) bool {
-	d.groupSelectionMutex.Lock()
-	defer d.groupSelectionMutex.Unlock()
-
-	minCurrentGroupCreationEndBlock := d.currentGroupSelectionStartBlock +
-		d.minGroupCreationDurationBlocks
-
-	shouldUpdate := d.currentGroupSelectionStartBlock == 0 ||
-		newGroupSelectionStartBlock > minCurrentGroupCreationEndBlock
-
-	if shouldUpdate {
-		d.currentGroupSelectionStartBlock = newGroupSelectionStartBlock
-		return true
-	}
-
-	return false
-}
-
```

Comment:
Quoting from the PR description:

> We removed the event deduplicator for the group selection events as it is no longer relevant for v2. We will need a new logic to handle v2 caveats. This was covered by a TODO in beacon.go

I am not sure if this is the case. As far as I remember, the deduplicator was mostly handling soft forks. We were receiving the same event twice but with a different block number: `n` and then `n+1`. The implementation is quite naive: it checks if the group selection had a chance to finish since the last emitted group selection started event and if not, it was considering the event as a duplicate of the last one.

Although this is naive, and we could improve it by, for example, waiting for N blocks, then checking the last N blocks if the event is included, I think the current implementation should do the work, even for v2.

Reply:
> As far as I remember, the deduplicator was mostly handling soft forks

Not only soft forks. The event subscription mechanism from contract bindings can replay the event when fetching past events. This happens more often than soft forks and results in an event having the same block.

>  I think the current implementation should do the work, even for v2.

This is something I'm not sure about. In v2 the minimum time between two `DkgStarted` events is `DKG protocol time + publication time + challenge period time`. In theory, we can expose `NotifyGroupCreationStarted` function from the event deduplicator to decide about group creation start. However, at the same time, we must be prepared for `DkgResultChallenged` events that actually restart the process and make the time between two `DkgStarted` events longer. The deduplicator must handle that. This is why I think we need to approach deduplication logic once we have challenges and approvals implemented together.

Reply:
> Not only soft forks. The event subscription mechanism from contract bindings can replay the event when fetching past events. This happens more often than soft forks and results in an event having the same block.

Yeah, that's a good point!

The time it takes the DKG protocol to produce the result is 72 blocks ([[1](https://github.com/keep-network/keep-core/blob/9f8d78ec3b7c91938db79dc9cc5882fec8ef49e9/solidity/random-beacon/contracts/libraries/BeaconDkg.sol#L117)], [[2](https://github.com/keep-network/keep-core/blob/9f8d78ec3b7c91938db79dc9cc5882fec8ef49e9/pkg/beacon/relay/gjkr/states.go#L13-L37)]). Subscription monitoring mechanism pulls 100 past blocks [by default](https://github.com/keep-network/keep-common/blob/58daa5270187dea35e7dd3642cc3541ecd45e0f6/pkg/chain/ethereum/ethutil/subscribe_opts.go#L12-L15) and although it could make sense for the relay request subscription, it does not make sense for DKG because after the first 6 blocks of inactivity, the game for the given member is over.

This solution is not ideal but at the same time, it's the simplest one. I am afraid that dealing with challenges will make this code pretty complicated.

I *think* that this should work:
1. Leave the deduplicator for DKG as-is, and removed just `currentGroupSelectionStartBlock`.
2. Make it clear `NotifyGroupSelectionStarted` filters out just for `minGroupCreationDurationBlocks` to the past.
3. When setting up a subscription for DKG, disable subscription monitoring loop. For 6 blocks, we would need to monitor pretty aggressively (every 10 seconds or so).

Does it make sense?

Reply:
We synced with @lukasz-zimnoch and came up with a better solution:
1. When setting up a subscription for DKG, disable subscription monitoring loop. 
2. Use a time-cache-based deduplicator and look at the DKG seed from the event.

Reply:
Here it is fbe3062b

Reply:
And added a TODO about disabling the monitoring loop 7c122d51